File: data-structure-typescript/src/list/array-list.ts
import type { List } from "./common/list.interface";

export class ArrayList<T> implements List<T> {
  private elements: Array<T>;
  private itemCount: number = 0;

  constructor(capacity: number = 10) {
    this.elements = new Array<T>(capacity);
    this.itemCount = 0;
  }

  add(element: T): void;
  add(index: number, element: T): void;

  add(...args: [T] | [number, T]): void {
    if (args.length !== 1 && args.length !== 2) {
      throw new Error("Invalid number of arguments for add method.");
    }

    if (args.length === 2) {
      const [index, element] = args;
      this.elements.splice(index, 0, element);
      this.itemCount++;
      return;
    }

    if (args.length === 1) {
      const [element] = args;
      this.elements[this.itemCount] = element;
      this.itemCount++;
    }
  }

  remove(index: number): T {
    if (index < 0 || index >= this.itemCount) {
      throw new Error("Index out of bounds.");
    }

    this.itemCount--;
    return this.elements.splice(index, 1)[0];
  }

  get(index: number): T {
    if (index < 0 || index >= this.itemCount) {
      throw new Error("Index out of bounds.");
    }

    return this.elements[index];
  }

  set(index: number, element: T): T {
    if (index < 0 || index >= this.itemCount) {
      throw new Error("Index out of bounds.");
    }

    this.elements[index] = element;

    return this.elements[index];
  }

  indexOf(element: T): number {
    if (this.itemCount === 0) {
      return -1;
    }

    return this.elements.indexOf(element);
  }

  clear(): void {
    this.elements = [];
    this.itemCount = 0;
  }

  isEmpty(): boolean {
    return this.itemCount === 0;
  }

  sizeOf(): number {
    return this.itemCount;
  }

  printAll(): void {
    const elementsToPrint: T[] = this.elements.slice(0, this.itemCount);

    console.log(elementsToPrint);
  }
}

File: data-structure-java/src/main/java/com/wisoft/io/datastructure/list/ArrayList.java
package com.wisoft.io.datastructure.list;

import com.wisoft.io.datastructure.list.common.ListInterface;

import java.util.Arrays;

public class ArrayList<T> implements ListInterface<T> {
    private Object[] array;
    private int size;
    private static final int DEFAULT_CAPACITY = 10;

    public ArrayList() {
        this.array = new Object[DEFAULT_CAPACITY];
        this.size = 0;
    }

    public ArrayList(int capacity) {
        if (capacity < 0) {
            throw new IllegalArgumentException("Illegal Capacity");
        }

        this.array = new Object[capacity];
        this.size = 0;
    }

    private boolean isFull() {
        return size == array.length;
    }

    @Override
    public void add(T e) {
        if (isFull()) {
            throw new IllegalStateException("ArrayList is full.");
        }

        array[size++] = e;
    }

    @Override
    public void add(int index, T element) {
        if (index < 0 || index > size) {
            throw new IndexOutOfBoundsException();
        }

        if (isFull()) {
            throw new IllegalStateException("ArrayList is full.");
        }

        shiftRight(index);

        array[index] = element;

        size++;
    }

    @Override
    public T remove(int index) {
        if (index < 0 || index >= size) {
            throw new IndexOutOfBoundsException();
        }

        T oldValue = (T) array[index];

        int numMoved = size - index - 1;
        if (numMoved > 0) {
            System.arraycopy(array, index + 1, array, index, numMoved);
        }

        array[--size] = null;

        return oldValue;
    }

    @Override
    public T get(int index) {
        if (index < 0 || index >= size) {
            throw new IndexOutOfBoundsException();
        }

        return (T) array[index];
    }

    @Override
    public T set(int index, T element) {
        if (index < 0 || index >= size) {
            throw new IndexOutOfBoundsException();
        }

        T oldValue = (T) array[index];

        array[index] = element;

        return oldValue;
    }

    @Override
    public int indexOf(T o) {
        if (o == null) {
            for (int i = 0; i < size; i++) {
                if (array[i] == null) {
                    return i;
                }
            }
        } else {
            for (int i = 0; i < size; i++) {
                if (o.equals(array[i])) {
                    return i;
                }
            }
        }

        return -1;
    }

    @Override
    public int size() {
        return size;
    }

    @Override
    public boolean isEmpty() {
        return size == 0;
    }

    @Override
    public void clear() {
        for (int i = 0; i < size; i++) {
            array[i] = null;
        }

        size = 0;
    }

    @Override
    public Object[] toArray() {
        return Arrays.copyOf(array, size);
    }

    @Override
    public String toString() {
        return Arrays.toString(Arrays.copyOf(array, size));
    }

    private void shiftRight(int index) {
        for (int i = size; i > index; i--) {
            array[i] = array[i - 1];
        }
    }

    public void printAll() {
        System.out.print("[");

        for (int i = 0; i < size; i++) {
            System.out.print(array[i]);
            if (i < size - 1) {
                System.out.print(", ");
            }
        }

        System.out.println("]");
    }
}